### PR TITLE
Implement settings menu and API

### LINKED ISSUE #32: Option to disable inputstreamhelper in settings + manual download cdm over GUI settings
Please work in 2 settings over GUI (maybe in inputstream.adaptive addon settings). It improves usability and avoids workaround over ssh:

1) "Option to disable inputstreamhelper" -> Not everybody has enabled automatic addon updates and updating cdm on an older inputstream.adaptive can cause incompatibility. Would be more user-friendly to have this option also over GUI.

2) Manual download cdm -> If users reject cdm update once (because of staying on older/incompatible inputstream.adaptive version, temporal reasons, by mistake..), there is no other chance for updating over Kodi GUI as to wait until inputstreamhelper asks again (this can take long time..).

### PR DESCRIPTION
**Pull request by dagwieers**
This PR includes:
- A Troubleshooting category with the option to disable the plugin
- Add an item to install Widevine CDM
- Add an item to remove Widevine CDM
- Add an API (addon.py) using script.module.routing

This fixes #32, #47, #52, #54 and #77

### CODE REVIEW

**Comment by dagwieers** [comment]
cc @horstle

**Comment by dagwieers** [comment]
cc @horstle @mediaminister @basrieter @emilsvennesson 

This is almost done. It adds and API to remove and install Widevine CDM.

In the future this can also be used to add an API for selecting a specific version or rolling back to a previous version of Widevine CDM. Please try it and provide feedback.

The nice thing about this is that we now have a proper addon-icon for users to access the settings directly. However I do expect that users will probably enter the settings through add-on settings references.

**Comment by dagwieers** [comment]
So I am not sure about the "disable" option, I still feel it should be "enabled" and "true" by default.
Now if you "enable" "disabled", the option to install/remove become disabled. Weird IMO.

**Comment by dagwieers** [comment]
@mediaminister If we have this added to inputstreamhelper, we can effectively remove our own handler from VRT NU. And everything falls into place ;-)

**Comment by horstle** [comment]
I think it would be better if it stayed "invisible". Most of the time users won't care about this and it would only be in their way, but if something fails, they would probably first look at the settings of the corresponding addon, so it's critical to get added there. Maybe at least in the inputstream.adaptive settings, since they are already referenced by some.

About the "disabled". If you really want to change this, I would suggest creating another setting "enabled" and hiding "disabled" from the settings page, then check for both settings and show a warning that "disabled" will be deprecated, if it is set. A few releases later you could then remove the "disabled" setting.

**Comment by horstle** [comment]
I think "disable" (or "enable") shouldn't be in the "Troubleshooting" category, but rather in "Configuration" (as you suggested in https://github.com/emilsvennesson/script.module.inputstreamhelper/issues/85#issue-482954477) or just be the very first one.

**Comment by dagwieers** [comment]
> I think it would be better if it stayed "invisible". 

We cannot make it invisible. The icon will appear. We can however make it do nothing.

> I think "disable" (or "enable") shouldn't be in the "Troubleshooting" category, but rather in "Configuration"

I disagree now. Everything in the settings are for troubleshooting.. The plugin should work out of the box and we should not expect people to have to use this.

That is why everything belongs together under Troubleshooting. And the upside is that we can disable everything if the plugin is disabled.

File: addon.xml [hunk 1]
```diff
@@ -2,10 +2,18 @@
 <addon id="script.module.inputstreamhelper" version="0.3.5" name="InputStream Helper" provider-name="emilsvennesson">
     <requires>
         <import addon="xbmc.python" version="2.25.0"/>
+        <import addon="script.module.inputstreamhelper" version="0.3.5" optional="true"/>
```
Comment (emilsvennesson) [inline addon.xml]: Why is this needed? :frowning:

**Comment by horstle** [comment]
> We cannot make it invisible. The icon will appear. We can however make it do nothing.
Cluttering the Addon section with an icon that does nothing would make no sense at all. If it has to be there, it might as well have a function.

**Comment by horstle** [comment]
> I disagree now. Everything in the settings are for troubleshooting.. The plugin should work out of the box and we should not expect people to have to use this.
> 
> That is why everything belongs together under Troubleshooting. And the upside is that we can disable everything if the plugin is disabled.

Once this is merged I'll rebase and fix #21 again. Should the setting for a custom path then be in the Troubleshooting section, too?

File: resources/settings.xml [hunk 2]
```diff
@@ -1,5 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <settings>
-   <setting label=30053 id="disabled" type="bool" value="" />
-   <setting id="last_update" value="" visible="false" />
+    <setting id="last_update" value="" visible="false" />
+    <setting id="version" value="" visible="false" />
+    <category label="30900"> <!-- Troubleshooting -->
+        <setting label="30903" help="30904" type="bool" id="disabled" value="false"/>
+        <setting label="30904" type="text" enable="false"/> <!-- disabled_warning -->
```
Comment (mediaminister) [inline resources/settings.xml]: The #30904 string is too long to show on one line in the settings dialog, it breaks at `playback ma...`

File: addon.xml (hunk 1 above)
Comment (dagwieers) [reply addon.xml]: Otherwise our "plugin" or "script" would not be able to import inputstreamhelper.
It looks silly, but it works.

File: resources/settings.xml (hunk 2 above)
Comment (dagwieers) [reply resources/settings.xml]: It worked for my setup at home and my celphone, but I was worried this, yes.
The real solution is to use the new v18 settings syntax which can show help statements.

**Comment by dagwieers** [comment]
> Perhaps we should split the API code into its own file? inputstreamhelper.py is starting to become quite complex, and it might be time to start separating some of the code.

I think we should, but can we postpone this until after v0.4.0 is released?
Can you create a ticket to discuss this?

**Comment by dagwieers** [comment]
Let's merge this now, and create PRs for the individual issues that were raised.
(I am off this weekend, so if this is merged, others can start building on this)

**Comment by mediaminister** [comment]
@emilsvennesson @dagwieers Removing Widevine CDM doesn't work on Windows. I have a fix for this, but I would like to clean up all Widevine CDM file operations.
But before I continue I would like to now why we use symlinks from the inpustreamhelper addon data dir and the Kodi CDM dir? Is this necessary? Is there a specific reason for this? Can I safely remove this to drastically simplify Widevine CDM file operations.

**Comment by horstle** [comment]
afaik the libwidevinecdm.so is stored in the inputstreamhelper data dir, but inputstream.adaptive (obviously) looks for it in the CDM dir. So a symlink seems the obvious solution. On Windows instead of a symlink the lib is actually copied instead of linked. That's probably why the removal doesn't work.

**Comment by mediaminister** [comment]
The removal didn't work because Windows uses `widevinecdm` and not `libwidevinecdm` as filename: https://github.com/peak3d/inputstream.adaptive/blob/980a42053c3fa0fa305d7349d4c74af50479f249/wvdecrypter/CMakeLists.txt#L41-L50

As a test I disabled inputstreamhelper and manually removed  inputstreamhelper cdm data dir.
DRM content just plays fine, as I expected because InputStream Adaptive always looks in the Kodi CDM dir:

https://github.com/peak3d/inputstream.adaptive/blob/80e4dbf147529f9edbd06a5b8b958c31666308da/CMakeLists.txt#L86-L90

My conclusion is that symlinking or copying `from` Kodi CDM dir to InputstreamHelper data dir is useless. It doesn't make sense to me.

**Comment by horstle** [comment]
> The removal didn't work because Windows uses `widevinecdm` and not `libwidevinecdm` as a filename:

Ok, didn't look at the code. Then that.

> As a test I disabled inputstreamhelper and manually removed inputstreamhelper cdm data dir.
> DRM content just plays fine, as I expected because InputStream Adaptive always looks in the Kodi CDM dir:

That's what I thought was the problem, but on linux that would have broken inputstream.adaptive, since the symlink in CDM dir would still exist, but not its target.

> My conclusion is that symlinking or copying **from** Kodi CDM dir **to** InputstreamHelper data dir is useless. It doesn't make sense to me.

I agree, copying from CDM to data would make no sense, but I couldn't find any code which would do that.
About the symlinks, a symlink in the CDM dir targeting the addon data dir (that's how I interpret "**from** Kodi CDM dir **to** InputstreamHelper data dir" in that case)  does make sense. However to create such a symlink with `os.symlink()` you have to specify the addon data dir first and then the CDM dir (a bit counter intuitive, but makes it consistent with `copy()`). Maybe that's where this issue comes from.

**Comment by mediaminister** [comment]
I wonder why we don't use the Kodi CDM folder to install the Widevine CDM. Symlinking from one directory to another directory makes the workflow more complicated. There must be reason for it, otherwise we should just use the Kodi CDM folder to install and remove Widevine CDM.

**Comment by emilsvennesson** [comment]
The reasoning behind this can be found in #1:

> @emilsvennesson Can't find the license where I read the part about extracting/copying it somewhere is forbidden. I did lots of research regarding available sources (legal ones) for ARM versions of widevine & I came to the conclusion that downloading the chromeos image & extracting it from there might be the best shot for not making Google angry.
I´d also asked friends at Mozilla & Google if there´s another way to get it for the development purposes only & they didn't know of any.

> My suggestion would be to use the ChromeOS image, symlink the libwidevine & delete everything      but the original path to the extracted version. Guess that is as close we can get to have a solution that might not violate the license.

_Originally posted by @asciidisco in https://github.com/emilsvennesson/script.module.inputstreamhelper/issues/1#issuecomment-329738013_

I don't think this is necessary though. I have yet to find anywhere in the Widevine license that states that it's forbidden to copy the binary (and we're already doing it on Windows), so this part of the code can definitely be simplified. It has just been there from the start and I've neglected to change it in favor of other things.

**Comment by mediaminister** [comment]
I completely agree with  @asciidisco 
A Kodi user already **copies** the software from a Google server to their hard disk when accepting the EULA. I think its better that this copy ends up in the right directory with as little as possible operations. In my opinion there is no problem as long as the software stays on the users (private) hard disk.
For instance, we should never upload copyrighted software to cloud services, even a (copyrighted) EULA or configuration file should never end up on Github.

**Comment by dagwieers** [comment]
So the required fix in **script.module.routing** (tamland/kodi-plugin-routing#16) has been accepted. We are now waiting for a new release.
